### PR TITLE
Add Conflict Grammar for Postgres

### PR DESCRIPTION
**Pull request by jmc-bbk**
### Brief summary of the change made

Added `ON CONFLICT` grammar to Postgres dialects.

Fixes https://github.com/sqlfluff/sqlfluff/issues/2904

### Are there any other side effects of this change that we should be aware of?

Not that I'm aware of.

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/dialects/dialect_postgres.py [hunk 1]
```diff
@@ -3217,6 +3217,98 @@ class AsAliasExpressionSegment(BaseSegment):
...
+                        ),
+                        # Sequence(
+                        #     Bracketed(
+                        #         Delimited(
+                        #             Ref("ColumnReferenceExpression")
+                        #         )
+                        #     ),
+                        #     Ref("EqualsSegment"),
+                        #     Bracketed(
+                        #         Ref("SubSelectExpression") # This isn't defined. Effectively an entire query.
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_postgres.py:3266]: Would `Bracketed(Ref("SelectableGrammar")),` work for you here? There's a few other instances where we do that.

Comment (jmc-bbk) [reply src/sqlfluff/dialects/dialect_postgres.py]: Thank you!

**Comment by jmc-bbk** [comment]
Letting the tests pass - but I need to:

- Merge main
- Remove a comment left for myself.

**Review by tunetheweb — APPROVED** [review]
LGTM!

**Pull request merged**